File: packages/Mosek/vBHcL/deps/src/mosek/8/tools/examples/python/case_portfolio_3.py
##
#  Copyright : Copyright (c) MOSEK ApS, Denmark. All rights reserved.
#
#  File :      case_portfolio_3.py
#
#  Purpose :   Implements a basic portfolio optimization model.
##
import mosek

def streamprinter(text):
    print("%s" % text),

if __name__ == '__main__':

    n = 3
    gamma = 0.05
    mu = [0.1073, 0.0737, 0.0627]
    GT = [[0.1667, 0.0232, 0.0013],
          [0.0000, 0.1033, -0.0022],
          [0.0000, 0.0000, 0.0338]]
    x0 = [0.0, 0.0, 0.0]
    w = 1.0
    m = [0.01, 0.01, 0.01]

    # This value has no significance.
    inf = 0.0

    with mosek.Env() as env:
        with env.Task(0, 0) as task:
            task.set_Stream(mosek.streamtype.log, streamprinter)

            rtemp = w
            for j in range(0, n):
                rtemp += x0[j]

            # Constraints.
            task.appendcons(1 + 9 * n)
            task.putconbound(0, mosek.boundkey.fx, rtemp, rtemp)
            task.putconname(0, "budget")

            task.putconboundlist(range(1 + 0, 1 + n), n *
                                 [mosek.boundkey.fx], n * [0.0], n * [0.0])
            for j in range(1, 1 + n):
                task.putconname(j, "GT[%d]" % j)

            task.putconboundlist(range(
                1 + n, 1 + 2 * n), n * [mosek.boundkey.lo], [-x0[j] for j in range(0, n)], n * [inf])
            for i in range(0, n):
                task.putconname(1 + n + i, "zabs1[%d]" % (1 + i))

            task.putconboundlist(range(1 + 2 * n, 1 + 3 * n),
                                 n * [mosek.boundkey.lo], x0, n * [inf])
            for i in range(0, n):
                task.putconname(1 + 2 * n + i, "zabs2[%d]" % (1 + i))

            task.putconboundlist(range(1 + 3 * n, 1 + 3 * n + 3 * n),
                                 3 * n * [mosek.boundkey.fx], 3 * n * [0.], 3 * n * [0.0])
            for i in range(0, n):
                for k in range(0, n):
                    task.putconname(1 + 3 * n + 3 * i + k,
                                    "f[%d,%d]" % (1 + i, 1 + k))

            task.putconboundlist(range(1 + 6 * n, 1 + 9 * n), 3 * n * [mosek.boundkey.fx],
                                 3 * [0.0, -1.0 / 8.0, 0.0], 3 * [0.0, -1.0 / 8.0, 0.0])
            for i in range(0, n):
                for k in range(0, n):
                    task.putconname(1 + 6 * n + 3 * i + k,
                                    "g[%d,%d]" % (1 + i, 1 + k))

            # Offset of variables into the API variable.
            offsetx = 0
            offsets = n
            offsett = n + 1
            offsetc = 2 * n + 1
            offsetv = 3 * n + 1
            offsetz = 4 * n + 1
            offsetf = 5 * n + 1
            offsetg = 8 * n + 1

            # Variables.
            task.appendvars(1 + 11 * n)

            # x variables.
            task.putclist(range(offsetx + 0, offsetx + n), mu)
            task.putaijlist(
                n * [0], range(offsetx + 0, offsetx + n), n * [1.0])
            for j in range(0, n):
                task.putaijlist(
                    n * [1 + j], range(offsetx + 0, offsetx + n), GT[j])
                task.putaij(1 + n + j, offsetx + j, -1.0)
                task.putaij(1 + 2 * n + j, offsetx + j, 1.0)

            task.putvarboundlist(
                range(offsetx + 0, offsetx + n), n * [mosek.boundkey.lo], n * [0.0], n * [inf])
            for j in range(0, n):
                task.putvarname(offsetx + j, "x[%d]" % (1 + j))

            # s variable.
            task.putvarbound(offsets + 0, mosek.boundkey.fx, gamma, gamma)
            task.putvarname(offsets + 0, "s")

            # t variables.
            task.putaijlist(range(1, n + 1), range(offsett +
                                                   0, offsett + n), n * [-1.0])
            task.putvarboundlist(range(offsett + 0, offsett + n),
                                 n * [mosek.boundkey.fr], n * [-inf], n * [inf])
            for j in range(0, n):
                task.putvarname(offsett + j, "t[%d]" % (1 + j))

            # c variables.
            task.putaijlist(n * [0], range(offsetc, offsetc + n), m)
            task.putaijlist(range(1 + 3 * n + 1, 1 + 6 * n + 1, 3),
                            range(offsetc, offsetc + n), n * [1.0])
            task.putvarboundlist(range(offsetc, offsetc + n),
                                 n * [mosek.boundkey.fr], n * [-inf], n * [inf])
            for j in range(0, n):
                task.putvarname(offsetc + j, "c[%d]" % (1 + j))

            # v variables.
            task.putaijlist(range(1 + 3 * n + 0, 1 + 6 * n + 0, 3),
                            range(offsetv, offsetv + n), n * [1.0])
            task.putaijlist(range(1 + 6 * n + 2, 1 + 9 * n + 2, 3),
                            range(offsetv, offsetv + n), n * [1.0])
            task.putvarboundlist(range(offsetv, offsetv + n),
                                 n * [mosek.boundkey.fr], n * [-inf], n * [inf])
            for j in range(0, n):
                task.putvarname(offsetv + j, "v[%d]" % (1 + j))

            # z variables.
            task.putaijlist(range(1 + 1 * n, 1 + 2 * n),
                            range(offsetz, offsetz + n), n * [1.0])
            task.putaijlist(range(1 + 2 * n, 1 + 3 * n),
                            range(offsetz, offsetz + n), n * [1.0])
            task.putaijlist(range(1 + 3 * n + 2, 1 + 6 * n + 2, 3),
                            range(offsetz, offsetz + n), n * [1.0])
            task.putaijlist(range(1 + 6 * n + 0, 1 + 9 * n + 0, 3),
                            range(offsetz, offsetz + n), n * [1.0])
            task.putvarboundlist(range(offsetz, offsetz + n),
                                 n * [mosek.boundkey.fr], n * [-inf], n * [inf])
            for j in range(0, n):
                task.putvarname(offsetz + j, "z[%d]" % (1 + j))

            # f variables.
            for j in range(0, n):
                for k in range(0, n):
                    task.putaij(1 + 3 * n + 3 * j + k,
                                offsetf + 3 * j + k, -1.0)
                    task.putvarbound(offsetf + 3 * j + k,
                                     mosek.boundkey.fr, -inf, inf)
                    task.putvarname(offsetf + 3 * j + k,
                                    "f[%d,%d]" % (1 + j, 1 + k))

            # g variables.
            for j in range(0, n):
                for k in range(0, n):
                    task.putaij(1 + 6 * n + 3 * j + k,
                                offsetg + 3 * j + k, -1.0)
                    task.putvarbound(offsetg + 3 * j + k,
                                     mosek.boundkey.fr, -inf, inf)
                    task.putvarname(offsetg + 3 * j + k,
                                    "g[%d,%d]" % (1 + j, 1 + k))

            task.appendcone(mosek.conetype.quad, 0.0, [
                            offsets] + list(range(offsett, offsett + n)))
            task.putconename(0, "stddev")

            for k in range(0, n):
                task.appendconeseq(mosek.conetype.rquad,
                                   0.0, 3, offsetf + 3 * k)
                task.putconename(1 + k, "f[%d]" % (1 + k))

            for k in range(0, n):
                task.appendconeseq(mosek.conetype.rquad,
                                   0.0, 3, offsetg + 3 * k)
                task.putconename(1 + n + k, "g[%d]" % (1 + k))

            task.putobjsense(mosek.objsense.maximize)

            # Turn all log output off.
            # task.putintparam(mosek.iparam.log,0)

            # Dump the problem to a human readable OPF file.
            #task.writedata("dump.opf")

            task.optimize()

            # Display the solution summary for quick inspection of results.
            task.solutionsummary(mosek.streamtype.msg)

            expret = 0.0
            x = [0.] * n
            task.getxxslice(mosek.soltype.itr, offsetx + 0, offsetx + n, x)
            for j in range(0, n):
                expret += mu[j] * x[j]

            stddev = [0.]
            task.getxxslice(mosek.soltype.itr, offsets +
                            0, offsets + 1, stddev)

            print("\nExpected return %e for gamma %e\n" % (expret, stddev[0]))

File: packages/Mosek/vBHcL/deps/src/mosek/8/tools/examples/dotnet/case_portfolio_3.cs
/*
  File : case_portfolio_3.cs

  Copyright : Copyright (c) MOSEK ApS, Denmark. All rights reserved.

  Description :  Implements a basic portfolio optimization model.
*/
using System;

namespace mosek.example {
  class msgclass : mosek.Stream
  {
    string prefix;
    public msgclass (string prfx)
    {
      prefix = prfx;
    }

    public override void streamCB (string msg)
    {
      Console.Write ("{0}{1}", prefix, msg);
    }
  }
  
  public class case_portfolio_3
  {
    public static void Main (String[] args)
    {
      const int n = 3;

      // Since the value infinity is never used, we define
      // 'infinity' symbolic purposes only
      double infinity = 0;
      double gamma = 0.05;
      double[]  mu = {0.1073,  0.0737,  0.0627};
      double[,] GT = {
        {0.1667,  0.0232,  0.0013},
        {0.0000,  0.1033, -0.0022},
        {0.0000,  0.0000,  0.0338}
      };
      double[] x0 = {0.0, 0.0, 0.0};
      double   w = 1.0;
      double[] m = {0.01, 0.01, 0.01};

      int offsetx = 0;
      int offsets = offsetx + n;
      int offsett = offsets + 1;
      int offsetc = offsett + n;
      int offsetv = offsetc + n;
      int offsetz = offsetv + n;
      int offsetf = offsetz + n;
      int offsetg = offsetf + 3 * n;

      int numvar = offsetg + 3 * n;

      int offset_con_budget = 0;
      int offset_con_gx_t = offset_con_budget + 1;
      int offset_con_abs1 = offset_con_gx_t + n;
      int offset_con_abs2 = offset_con_abs1 + n;
      int offset_con_f =  offset_con_abs2 + n;
      int offset_con_g =  offset_con_f + 3 * n;

      int numcon = 1 + 3 * n + 2 * 3 * n;


      // Make mosek environment.
      using (mosek.Env env = new mosek.Env())
      {
        // Create a task object.
        using (mosek.Task task = new mosek.Task(env, 0, 0))
        {
          // Directs the log task stream to the user specified
          // method msgclass.streamCB
          task.set_Stream(mosek.streamtype.log, new msgclass(""));


          //Set up constraint bounds, names and variable coefficients
          task.appendcons(numcon);
          for (int i = 0; i < n; ++i)
          {
            w += x0[i];
            task.putconbound(offset_con_gx_t + i, mosek.boundkey.fx, 0.0, 0.0);
            task.putconname(offset_con_gx_t + i, "GT[" + (i + 1) + "]");

            task.putconbound(offset_con_abs1 + i, mosek.boundkey.lo, -x0[i], infinity);
            task.putconname(offset_con_abs1 + i, "zabs1[" + (i + 1) + "]");

            task.putconbound(offset_con_abs2 + i, mosek.boundkey.lo, x0[i], infinity);
            task.putconname(offset_con_abs2 + i, "zabs2[" + (i + 1) + "]");

            for (int j = 0; j < 3; ++j)
            {
              task.putconbound(offset_con_f + 3 * i + j, mosek.boundkey.fx, 0.0, 0.0);
              task.putconname(offset_con_f + 3 * i + j, "f[" + (i + 1) + "," + (j + 1) + "]");

              task.putconbound(offset_con_g + 3 * i + j, mosek.boundkey.fx, 0.0, 0.0);
              task.putconname(offset_con_g + 3 * i + j, "g[" + (i + 1) + "," + (j + 1) + "]");
            }

            task.putconbound(offset_con_g + 3 * i + 1, mosek.boundkey.fx, -1.0 / 8.0, -1.0 / 8.0);

          }
          // e x = w + e x0
          task.putconbound(offset_con_budget, mosek.boundkey.fx, w, w);
          task.putconname(offset_con_budget, "budget");

          //Variables.
          task.appendvars(numvar);

          //the objective function coefficients
          int[] xindx = { offsetx + 0, offsetx + 1, offsetx + 2 };
          task.putclist(xindx, mu);

          double[] one_m_one = { 1.0, -1.0 };
          double[] one_one = { 1.0, 1.0 };

          //set up variable bounds and names
          for (int i = 0; i < n; ++i)
          {
            task.putvarbound(offsetx + i, mosek.boundkey.lo, 0.0, infinity);
            task.putvarbound(offsett + i, mosek.boundkey.fr, infinity, infinity);
            task.putvarbound(offsetc + i, mosek.boundkey.fr, infinity, infinity);
            task.putvarbound(offsetz + i, mosek.boundkey.fr, infinity, infinity);
            task.putvarbound(offsetv + i, mosek.boundkey.fr, infinity, infinity);
            for (int j = 0; j < 3; ++j)
            {
              task.putvarbound(offsetf + j + i * 3, mosek.boundkey.fr, infinity, infinity);
              task.putvarbound(offsetg + j + i * 3, mosek.boundkey.fr, infinity, infinity);
            }
            task.putvarname(offsetx + i, "x[" + (i + 1) + "]");
            task.putvarname(offsett + i, "t[" + (i + 1) + "]");
            task.putvarname(offsetc + i, "c[" + (i + 1) + "]");
            task.putvarname(offsetz + i, "z[" + (i + 1) + "]");
            task.putvarname(offsetv + i, "v[" + (i + 1) + "]");
            for (int j = 0; j < 3; ++j)
            {
              task.putvarname(offsetf + j + i * 3, "f[" + (i + 1) + "," + (j + 1) + "]");
              task.putvarname(offsetg + j + i * 3, "g[" + (i + 1) + "," + (j + 1) + "]");
            }

            for (int j = i; j < n; ++j)
              task.putaij(offset_con_gx_t + i, j, GT[i, j]);

            task.putaij(offset_con_gx_t + i, offsett + i, -1.0);

            task.putaij(offset_con_budget, offsetx + i, 1.0);
            task.putaij(offset_con_budget, offsetc + i, m[i]);

            // z_j - x_j >= -x0_j
            int[] indx1 = { offsetz + i, offsetx + i };
            task.putarow(offset_con_abs1 + i, indx1, one_m_one);
            // z_j + x_j >= +x0_j
            int[] indx2 = { offsetz + i, offsetx + i };
            task.putarow(offset_con_abs2 + i, indx2, one_one);

            int[] indxf1 = { offsetv + i, offsetf + i * 3 };
            task.putarow(offset_con_f + 3 * i, indxf1, one_m_one);
            int[] indxf2 = { offsetc + i, offsetf + i * 3 + 1 };
            task.putarow(offset_con_f + 1 + 3 * i, indxf2, one_m_one);
            int[] indxf3 = { offsetz + i, offsetf + i * 3 + 2 };
            task.putarow(offset_con_f + 2 + 3 * i, indxf3, one_m_one);

            int[] indxg1 = { offsetz + i, offsetg + i * 3 };
            task.putarow(offset_con_g + 3 * i, indxg1, one_m_one);

            task.putaij(offset_con_g + 3 * i + 1, offsetg + i * 3 + 1, -1.0);

            int[] indxg3 = { offsetv + i, offsetg + i * 3 + 2 };
            task.putarow(offset_con_g + 3 * i + 2, indxg3, one_m_one);
          }
          task.putvarbound(offsets, mosek.boundkey.fx, gamma, gamma);
          task.putvarname(offsets, "s");

          //Cones.
          int conecount = 0;

          int[] csub = { offsets, offsett + 0, offsett + 1, offsett + 2 };
          task.appendcone(mosek.conetype.quad, 0.0, csub);
          task.putconename(conecount, "stddev");
          ++conecount;

          for (int j = 0; j < n; ++j, ++conecount)
          {
            int[] coneindx = { offsetf + j * 3, offsetf + j * 3 + 1, offsetf + j * 3 + 2 };
            task.appendcone(mosek.conetype.rquad, 0.0, coneindx);
            task.putconename(conecount, "f[" + (j + 1) + "]");
          }

          for (int j = 0; j < n; ++j, ++conecount)
          {
            int[] coneindx = { offsetg + j * 3, offsetg + j * 3 + 1, offsetg + j * 3 + 2 };
            task.appendcone(mosek.conetype.rquad, 0.0, coneindx);
            task.putconename(conecount, "g[" + (j + 1) + "]");
          }
          /* A maximization problem */
          task.putobjsense(mosek.objsense.maximize);

          //Turn all log output off.
          //task.putintparam(mosek.iparam.log,0);

          //task.writedata("dump.opf");
          /* Solve the problem */
          task.optimize();

          task.solutionsummary(mosek.streamtype.log);

          double expret = 0.0;
          double[] xx = new double[numvar];

          task.getxx(mosek.soltype.itr, xx);

          for (int j = 0; j < n; ++j)
            expret += mu[j] * xx[j + offsetx];

          Console.WriteLine("Expected return {0:E6} for gamma {1:E6}\n\n", expret, xx[offsets]);
        }
      }
    }
  }
}